examples/OpenGL45/arrayReductionByAddition_version4.shadertrap

```diff
@@ -0,0 +1,67 @@
+# Copyright 2021 Shiyu Liu
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+GL 4.5
+
+CREATE_BUFFER array1 SIZE_BYTES 1024 INIT_VALUES uint 0 1 2 3 4 5 6 7 8 9 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 25 26 27 28 29 30 31 32 33 34 35 36 37 38 39 40 41 42 43 44 45 46 47 48 49 50 51 52 53 54 55 56 57 58 59 60 61 62 63 64 65 66 67 68 69 70 71 72 73 74 75 76 77 78 79 80 81 82 83 84 85 86 87 88 89 90 91 92 93 94 95 96 97 98 99 100 101 102 103 104 105 106 107 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232 233 234 235 236 237 238 239 240 241 242 243 244 245 246 247 248 249 250 251 252 253 254 255
+
+CREATE_BUFFER result SIZE_BYTES 4 INIT_VALUES uint 0
+CREATE_BUFFER expected SIZE_BYTES 4 INIT_VALUES uint 32640
+
+BIND_SHADER_STORAGE_BUFFER BUFFER array1 BINDING 0
+BIND_SHADER_STORAGE_BUFFER BUFFER result BINDING 1
+
+DECLARE_SHADER shader KIND COMPUTE
+#version 450
+
+layout(local_size_x=256, local_size_y=1, local_size_z=1) in;
+
+layout(std430, binding = 0) buffer array1 {
+  uint[256] data1;
+}; 
+layout(std430, binding = 1) buffer result{
+  uint res;
+};
+
+shared uint shared_memory[256];
+
+void main() {
+  // Same as version 3 but we are using shared memory instead of global memory.
+  uint id = uint(gl_LocalInvocationID.x);
+  uint depth = findMSB(data1.length());
+  
+  // Data is copied over to shared memory by 256 threads in a work group.
+  shared_memory[id] = data1[id];
+  barrier();
+  
+  for(uint i = 1; i <= depth; ++i) {
+  	if(id % (1 << i) == 0) shared_memory[id] += shared_memory[id + (1<<(i-1))];
+  	barrier();  	 
+  }
+  
+  // Update global buffer with shared memory. 
+  data1[id] = shared_memory[id];
+  
+  res = data1[0];
```

Comment:
This is an example of a genuine data race. Consider the case where:

- invocation 1 reads data1[0] right before invocation 0 writes to it at line 54
- invocation 0 then writes data1[0], reads from it and updates res
- invocation 0 then updates res to the stale value that it read from data1[0] before invocation 0 had written to it

Thus you should guard the final statement so that only invocation 0 executes it.

Also: it would be fine to eliminate the write to `data1` and just have invocation 0 do:

`res = shared_memory[0];`

However, you could keep the write to `data1` if you want to do result checking (see below).

Reply:
done.